queue.cpp

```diff
@@ -0,0 +1,99 @@
+#include <iostream>
+#include <stack>
+#include <algorithm>
+#include <functional>
+
+using namespace std;
+
+template<typename T, typename F = function<T(T, T)>>
+class AggQueue {
+private:
+    stack<pair<T, T>> inputStack, outputStack;
+    F func;
+
+    void pushWithAgg(stack<pair<T, T>>& stack, T value) {
+        T currentAgg = stack.empty() ? value : func(value, stack.top().second);
+        stack.push({value, currentAgg});
+    }
+
+    void transferIfNeeded() {
+        if (outputStack.empty()) {
+            while (!inputStack.empty()) {
+                T element = inputStack.top().first;
+                pushWithAgg(outputStack, element);
```

Comment:
Is it needed to call functor twice?

Reply:
Yes, because inputStack and outputStack will store different aggregate values due to the different order of adding values.